### PR TITLE
initial checkin, expanded rpm packaging for openresty

### PR DESCRIPTION
**Pull request by zxcvbn4038**
@agentzh As discussed on the Google group -

This is a starting point for an expanded Redhat packaging for Openresty. This is functional but lacks polish and am hoping that someone who knows more about making RPMs can improve on what I've done.

The main problem is that a lot of Redhat systems in use today have hard dependencies on older versions of zlib, openssl, pcre, etc. then what is needed for Openresty w/ http2 support. So the idea here is to provide independent libraries for Openresty to use, without interfering with the native libraries provided by redhat, and use the rpath linker option to ensure openresty uses those libraries instead of the system libraries.

This change provides:

openresty-zlib - provides zlib 1.2.8, no dependencies
openresty-openssl - provides openssl 1.0.2h, dependency on openresy-zlib
openresty-pcre - provides pcre 8.3.9, no dependencies
openresty - provides openresty, dependency on openresty-{zlib,openssl,pcre}

Known issues:

- white space is all over the place
- pcre sets up its shared libraries different then other libraries, am concerned that openresty-pcre-devel doesn't work like it should (but it does produce a valid end result)
- @agentzh mentioned doing the same thing with luajit library, but I didn't include that in this PR

Things that can use polish:
- Are the macros used appropriately? Maybe there are better ways I don't know about
- I'm building static libraries but probably don't have to because we don't use them
- I'm building pcre16.so and pcre32.so also, but I think only pcre is used by openresty
- pcre tools probably don't need to be there
- I included the openssl command line client because you need that when testing/debugging features that the native openssl library doesn't provide
- I'm building man pages but probably don't have to

### CODE REVIEW

File: rpm/SOURCES/patch_1658.diff [hunk 1]
```diff
@@ -0,0 +1,15 @@
+--- code/trunk/pcregrep.c	2016/06/14 10:27:41	1657
```
Comment (agentzh) [inline rpm/SOURCES/patch_1658.diff]: Please exclude this PCRE patch. We don't need it at all.

File: rpm/SPECS/openresty-openssl.spec [hunk 2]
```diff
@@ -14,10 +14,14 @@ Patch0:             https://raw.githubusercontent.com/openresty/openresty/master
 
 BuildRoot:          %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
 
-BuildRequires:      gcc, make
+BuildRequires:      gcc, make, 
+BuildRequires:			openresty-zlib-devel >= 1.2.8
+Requires:						openresty-zlib >= 1.2.8
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Please fix the spaces here to make the columns aligned up.

File: rpm/SPECS/openresty-openssl.spec [hunk 3]
```diff
@@ -35,26 +39,22 @@ Provides C header and static library for OpenResty's OpenSSL library.
 
 %prep
 %setup -q -n openssl-%{version}
-
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Why this change? The blank line is there on purpose.

File: rpm/SPECS/openresty-openssl.spec [hunk 4]
```diff
@@ -35,26 +39,22 @@ Provides C header and static library for OpenResty's OpenSSL library.
...
 %patch0 -p1
 
 
 %build
-./config --prefix=%{openssl_prefix} \
-    no-threads \
-    shared -g
-
+./config --openssldir=%{_prefix} no-threads shared zlib -g \
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Why removing `--prefix=PATH` and adding this `--openssldir`?

File: rpm/SPECS/openresty-openssl.spec [hunk 5]
```diff
@@ -35,26 +39,22 @@ Provides C header and static library for OpenResty's OpenSSL library.
...
+-I%{openresty_prefix}/include \
+-L%{openresty_prefix}/%{_lib} \
+-Wl,-rpath,%{openresty_prefix}/%{_lib}
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: The zlib headers and libs should be installed into the prefix `%{openresty_prefix}/zlib/`.

File: rpm/SPECS/openresty-openssl.spec [hunk 6]
```diff
@@ -35,26 +39,22 @@ Provides C header and static library for OpenResty's OpenSSL library.
...
 
 
 %install
 make install_sw INSTALL_PREFIX=%{buildroot}
 
-chmod +w %{buildroot}%{openssl_prefix}/lib/*.so
-chmod +w %{buildroot}%{openssl_prefix}/lib/*/*.so
-
-rm -rf %{buildroot}%{openssl_prefix}/bin
-rm -rf %{buildroot}%{openssl_prefix}/ssl
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Why removing these `rm` commands?

File: rpm/SPECS/openresty-openssl.spec [hunk 7]
```diff
@@ -67,18 +67,21 @@ rm -rf %{buildroot}
 %files
 %defattr(-,root,root,-)
 
-%attr(0755,root,root) %{openssl_prefix}/lib/*.so*
-%attr(0755,root,root) %{openssl_prefix}/lib/*/*.so*
-
+%attr(0755,root,root) %{openresty_prefix}/bin/openssl
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Okay, I see you want to include the `openssl` command-line utility. That's fair. But I think we should install it into `%{openssl_prefix}/bin/` instead.

File: rpm/SPECS/openresty-openssl.spec [hunk 8]
```diff
@@ -67,18 +67,21 @@ rm -rf %{buildroot}
...
+%attr(0755,root,root) %{openresty_prefix}/%{_lib}/*.so*
+%attr(0755,root,root) %{openresty_prefix}/%{_lib}/*/*.so*
+%exclude %{openresty_prefix}/bin/c_rehash
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: No, we don't need this.

File: rpm/SPECS/openresty-openssl.spec [hunk 9]
```diff
@@ -67,18 +67,21 @@ rm -rf %{buildroot}
...
+%exclude %{openresty_prefix}/misc
+%exclude %{openresty_prefix}/openssl.cnf
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Nor these.

File: rpm/SPECS/openresty-openssl.spec [hunk 10]
```diff
@@ -67,18 +67,21 @@ rm -rf %{buildroot}
...
 
 %files devel
 %defattr(-,root,root,-)
 
-%{openssl_prefix}/include/*
-%{openssl_prefix}/lib/pkgconfig/*
-%attr(0755,root,root) %{openssl_prefix}/lib/*.a
+%{openresty_prefix}/include/*
+%{openresty_prefix}/%{_lib}/pkgconfig/*
+%attr(0755,root,root) %{openresty_prefix}/%{_lib}/*.a
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: This openssl library should be installed into the `%{openssl_prefix}` location. Please do not change that.

File: rpm/SPECS/openresty-pcre.spec [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+License:    BSD
+URL:        http://www.pcre.org/
+Source0:    ftp://ftp.csx.cam.ac.uk/pub/software/programming/pcre/pcre-%{version}.tar.bz2
+
+# http://vcs.pcre.org/pcre/code/trunk/pcregrep.c?view=patch&r1=1658&r2=1657&pathrev=1658
+Patch0:     patch_1658.diff
+
+BuildRequires:  readline-devel
+BuildRequires:  autoconf
+BuildRequires:  automake
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: I don't think we need the autotools since we can invoke the `./configure` script directly.

File: rpm/SPECS/openresty-pcre.spec [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+Summary:    Perl-compatible regular expression library for OpenResty
+
+Group:      System Environment/Libraries
+
+License:    BSD
+URL:        http://www.pcre.org/
+Source0:    ftp://ftp.csx.cam.ac.uk/pub/software/programming/pcre/pcre-%{version}.tar.bz2
+
+# http://vcs.pcre.org/pcre/code/trunk/pcregrep.c?view=patch&r1=1658&r2=1657&pathrev=1658
+Patch0:     patch_1658.diff
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Let's get rid of this patch. We do not need to bundle `pcregrep` at all.

File: rpm/SPECS/openresty-pcre.spec [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+BuildRequires:  coreutils
+BuildRequires:  gcc
+BuildRequires:  glibc-common
+BuildRequires:  libtool
+BuildRequires:  make
+# perl not used because config.h.generic is pregenerated
+# Tests:
+BuildRequires:  bash
+BuildRequires:  diffutils
+BuildRequires:  grep
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: I don't think we need `diffutils` or `grep` here?

File: rpm/SPECS/openresty-pcre.spec [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+%description tools
+Utilities demonstrating PCRE capabilities like pcregrep or pcretest.
+
+%prep
+%setup -q -n pcre-%{version}
+%patch0 -p2
+
+%build
+LDFLAGS="-Wl,-rpath,%{openresty_prefix}/%{_lib}" ./configure \
+  --prefix=%{openresty_prefix} \
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: The PCRE library should be installed into the prefix `%{openresty_prefix}/pcre` (or better, just `%{pcre_prefix}` which should be defined previously in this spec file).

File: rpm/SPECS/openresty-pcre.spec [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+  --libdir=%{openresty_prefix}/%{_lib} \
+  --enable-jit \
+  --enable-pcretest-libreadline --enable-utf --enable-unicode-properties \
+  --enable-pcre8 --enable-pcre16 --enable-pcre32
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Let's get rid of `pcre16` and `pcre32` here since we do not use them at all.

File: rpm/SPECS/openresty-pcre.spec [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+%prep
+%setup -q -n pcre-%{version}
+%patch0 -p2
+
+%build
+LDFLAGS="-Wl,-rpath,%{openresty_prefix}/%{_lib}" ./configure \
+  --prefix=%{openresty_prefix} \
+  --libdir=%{openresty_prefix}/%{_lib} \
+  --enable-jit \
+  --enable-pcretest-libreadline --enable-utf --enable-unicode-properties \
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: `pcretest-libreadline` sholud be disabled.

File: rpm/SPECS/openresty-pcre.spec [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+  --enable-pcre8 --enable-pcre16 --enable-pcre32
+make %{?_smp_mflags}
+
+%install
+make install DESTDIR=%{buildroot}
+mv -f %{buildroot}/%{openresty_prefix}/share/man %{buildroot}/%{openresty_prefix}/man
+
+%check
+make %{?_smp_mflags} check VERBOSE=yes
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: We could remove this to speed up the build, I think.

File: rpm/SPECS/openresty-pcre.spec [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+
+%post -p /sbin/ldconfig
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: No no no, do not run `ldconfig` in this package since we do not want other programs built on this machine to use our OpenSSL library.

File: rpm/SPECS/openresty-pcre.spec [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+
+%postun -p /sbin/ldconfig
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Ditto.

File: rpm/SPECS/openresty-pcre.spec [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+%exclude %{openresty_prefix}/%{_lib}/*.la
+%exclude %{openresty_prefix}/share
+
+%files devel
+%defattr(-,root,root,-)
+#FIX%{openresty_prefix}/%{_lib}/*.so
+%{openresty_prefix}/%{_lib}/pkgconfig/*
+%{openresty_prefix}/include/*.h
+%{openresty_prefix}/man/man1/pcre-config.*
+%{openresty_prefix}/man/man3/*
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Let's get rid of the docs.

File: rpm/SPECS/openresty-pcre.spec [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+%files
+%defattr(-,root,root,-)
+%{openresty_prefix}/%{_lib}/*.so.*
+%exclude %{openresty_prefix}/%{_lib}/*.la
+%exclude %{openresty_prefix}/share
+
+%files devel
+%defattr(-,root,root,-)
+#FIX%{openresty_prefix}/%{_lib}/*.so
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Let's remove this line.

File: rpm/SPECS/openresty-pcre.spec [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+%{openresty_prefix}/%{_lib}/pkgconfig/*
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: We can exclude the `pkgconfig` directory I think.

File: rpm/SPECS/openresty-pcre.spec [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+%{openresty_prefix}/include/*.h
+%{openresty_prefix}/man/man1/pcre-config.*
+%{openresty_prefix}/man/man3/*
+%{openresty_prefix}/bin/pcre-config
+%exclude %{openresty_prefix}/%{_lib}/*.so
+
+%files static
+%defattr(-,root,root,-)
+%{openresty_prefix}/%{_lib}/*.a
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Let's avoid a `static` sub-package. Just merge it into the `devel` package.

File: rpm/SPECS/openresty-pcre.spec [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+
+%files tools
+%defattr(-,root,root,-)
+%{openresty_prefix}/bin/pcregrep
+%{openresty_prefix}/bin/pcretest
+%{openresty_prefix}/man/man1/pcregrep.*
+%{openresty_prefix}/man/man1/pcretest.*
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Let's get rid of the `tools` sub-package altogether.

File: rpm/SPECS/openresty.spec [hunk 25]
```diff
@@ -165,7 +164,7 @@ export QA_RPATHS=$[ 0x0002 ]
 
 
 %clean
-rm -rf %{buildroot}
+#rm -rf %{buildroot}
```
Comment (agentzh) [inline rpm/SPECS/openresty.spec]: Why comment this line out?

File: rpm/SPECS/openresty.spec [hunk 26]
```diff
@@ -108,11 +110,8 @@ services, and dynamic web gateways.
 
 %build
 ./configure \
-    --with-cc-opt="-I%{orprefix}/openssl/include" \
-    --with-ld-opt="-L%{orprefix}/openssl/lib -Wl,-rpath,%{orprefix}/openssl/lib" \
-    --with-zlib=../zlib-%{zlib_version} \
-    --with-pcre=../pcre-%{pcre_version} \
-    --with-pcre-opt="-DSUPPORT_UTF" \
+    --with-cc-opt="-I%{orprefix}/include -I%{orprefix}/openssl/include" \
+    --with-ld-opt="-L%{orprefix}/%{_lib} -Wl,-rpath,%{orprefix}/%{_lib}" \
```
Comment (agentzh) [inline rpm/SPECS/openresty.spec]: These paths should be updated for the right prefixes of zlib, pcre, and openssl (they should have their own sub-directories under `%{orprefix}/` instead of flattening directly.

File: rpm/SPECS/openresty-zlib.spec [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+%{openresty_prefix}/include/zlib.h
+%{openresty_prefix}/include/zconf.h
+%{openresty_prefix}/man/man3/zlib.3*
+
+%files static
+%defattr(-,root,root,-)
+%attr(0755,root,root) %{openresty_prefix}/%{_lib}/libz.a
+
+%changelog
+
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: Need a stub change log here.

File: rpm/SPECS/openresty-zlib.spec [hunk 28]
```diff
@@ -0,0 +1,88 @@
+Name:						openresty-zlib
+Version:				1.2.8
+Release:				1%{?dist}
+Summary:				The compression and decompression library for OpenResty
+
+Group:					System Environment/Libraries
+
+# /contrib/dotzlib/ have Boost license
+License:				zlib and Boost
+URL:						http://www.zlib.net/
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: Please fix the spaces to make the 2nd column aligned up. Thanks!

File: rpm/SPECS/openresty-zlib.spec [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+Source0:				http://www.zlib.net/zlib-%{version}.tar.xz
+
+BuildRoot:			%{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
+
+BuildRequires:	automake, autoconf, libtool
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: We can get rid of these I think.

File: rpm/SPECS/openresty-zlib.spec [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+
+%define openresty_prefix /usr/local/openresty
+
+# Do not check openresty files for provides, our internals are not for others.
+AutoReqProv: no
+
+%description
+Zlib is a general-purpose, patent-free, lossless data compression
+library which is used by many different programs.
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: We should make it clear that this package is for OpenResty only.

File: rpm/SPECS/openresty-zlib.spec [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+
+%package devel
+
+Summary:				Header files and libraries for Zlib development
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: Mention OpenResty only here.

File: rpm/SPECS/openresty-zlib.spec [hunk 32]
```diff
@@ -0,0 +1,88 @@
...
+Group:					Development/Libraries
+Requires:				%{name} = %{version}-%{release}
+
+%description devel
+The zlib-devel package contains the header files and libraries needed
+to develop programs that use the zlib compression and decompression
+library.
+
+%package static
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: Let's merge the static package into devel.

File: rpm/SPECS/openresty-zlib.spec [hunk 33]
```diff
@@ -0,0 +1,88 @@
...
+
+%check
+make test
+
+%install
+make install DESTDIR=%{buildroot}
+mv -f %{buildroot}/%{openresty_prefix}/share/man %{buildroot}/%{openresty_prefix}/man
+rm -f %{buildroot}/%{openresty_prefix}/%{_lib}/*.la
+
+%post -p /sbin/ldconfig
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: Again: no `ldconfig` (please change other places accordingly).

File: rpm/SPECS/openresty-zlib.spec [hunk 34]
```diff
@@ -0,0 +1,88 @@
...
+  --includedir=%{openresty_prefix}/include \
+  --prefix=%{openresty_prefix}
+make %{?_smp_mflags}
+
+%check
+make test
+
+%install
+make install DESTDIR=%{buildroot}
+mv -f %{buildroot}/%{openresty_prefix}/share/man %{buildroot}/%{openresty_prefix}/man
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: No does. Let's remove them.

File: rpm/SPECS/openresty-openssl.spec (hunk 5 above)
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Style: need indentation for these continued lines to make them look better.

File: rpm/SPECS/openresty-pcre.spec [hunk 35]
```diff
@@ -0,0 +1,119 @@
...
+
+%clean
+# test
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Please add the line `rm -rf %{buildroot}` here. Thank you.

**Comment by agentzh** [comment]
@zxcvbn4038 Thank you for your contribution! I've just done one pass of review and left some comments. Will you take care of them? Many thanks!

File: rpm/SPECS/openresty-pcre.spec [hunk 36]
```diff
@@ -0,0 +1,119 @@
...
+Group:      System Environment/Libraries
+
+License:    BSD
+URL:        http://www.pcre.org/
+Source0:    ftp://ftp.csx.cam.ac.uk/pub/software/programming/pcre/pcre-%{version}.tar.bz2
+
+# http://vcs.pcre.org/pcre/code/trunk/pcregrep.c?view=patch&r1=1658&r2=1657&pathrev=1658
+Patch0:     patch_1658.diff
+
+BuildRequires:  readline-devel
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: No, no readline.

**Comment by zxcvbn4038** [comment]
@agentzh I don't disagree with any of your comments and will revise the PR like you've asked. Look for something in a few days. I'll try to add in luajit also - you had mentioned wanting that on the Google group.

File: rpm/SPECS/openresty-openssl.spec (hunk 4 above)
Comment (zxcvbn4038) [inline rpm/SPECS/openresty-openssl.spec]: @agentzh Looking at https://wiki.openssl.org/index.php/Compilation_and_Installation#PREFIX_and_OPENSSLDIR I interpret that --openssldir is preferable.

Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: @zxcvbn4038 OK, we can use `--openssldir`, but please point the DIR to the original `%{openssl_prefix}` location. I just wonder why you changed it to `%{_prefix}` in your branch.

**Comment by zxcvbn4038** [comment]
@agentzh There is still more to do but I think this is ready for a second round of feedback.

- I didn't do luajit as a package yet
- the debug/valgrind versions of the spec files need to be updated
- I'd like to get rid of this lib64 stuff, openssl is doing something internal to guess which should be, I'd like it to guess "lib" every time and all the other packages to use lib also

File: rpm/SPECS/openresty-openssl.spec [hunk 37]
```diff
@@ -1,23 +1,28 @@
-Name:               openresty-openssl
-Version:            1.0.2h
-Release:            4%{?dist}
-Summary:            OpenSSL library for OpenResty
+Name:          openresty-openssl
+Version:       1.0.2h
+Release:       5%{?dist}
+Summary:       OpenSSL library for OpenResty
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Could you preserve my original space count? Let's keep the changes minimal.

Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: And it's also easier for code review.

File: rpm/SPECS/openresty-openssl.spec [hunk 38]
```diff
@@ -1,23 +1,28 @@
...
-BuildRoot:          %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
+BuildRoot:     %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
 
-BuildRequires:      gcc, make
+BuildRequires: gcc, make, 
+BuildRequires: openresty-zlib-devel >= 1.2.8
+Requires:      openresty-zlib >= 1.2.8
 
+%define openssl_prefix /usr/local/openresty/openssl
+%define zlib_prefix /usr/local/openresty/zlib
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: The 2nd column should be aligned.

File: rpm/SPECS/openresty-openssl.spec [hunk 39]
```diff
@@ -66,19 +77,15 @@ rm -rf %{buildroot}
 
 %files
 %defattr(-,root,root,-)
-
-%attr(0755,root,root) %{openssl_prefix}/lib/*.so*
-%attr(0755,root,root) %{openssl_prefix}/lib/*/*.so*
-
+%attr(0755,root,root) %{openssl_prefix}/bin/openssl
+%attr(0755,root,root) %{openssl_prefix}/%{_lib}/*.so*
+%attr(0755,root,root) %{openssl_prefix}/%{_lib}/*/*.so*
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Why not just use `lib` here?

File: rpm/SPECS/openresty.spec [hunk 40]
```diff
@@ -1,37 +1,38 @@
-Name:           openresty
-Version:        1.9.15.1
-Release:        16%{?dist}
-Summary:        OpenResty, scalable web platform by extending NGINX with Lua
+Name:          openresty
+Version:       1.9.15.1
+Release:       16%{?dist}
+Summary:       OpenResty, scalable web platform by extending NGINX with Lua
```
Comment (agentzh) [inline rpm/SPECS/openresty.spec]: Please do not introduce unnecessary changes like above.

File: rpm/SPECS/openresty.spec [hunk 41]
```diff
@@ -54,12 +55,12 @@ a single box.
 
 %package resty
 
-Summary:        OpenResty command-line utility, resty
-Group:          Development/Tools
-Requires:       perl, openresty
+Summary:       OpenResty command-line utility, resty
+Group:         Development/Tools
+Requires:      perl, openresty
```
Comment (agentzh) [inline rpm/SPECS/openresty.spec]: Ditto.

File: rpm/SPECS/openresty.spec [hunk 42]
```diff
@@ -164,10 +160,6 @@ mkdir -p %{buildroot}/etc/init.d
 export QA_RPATHS=$[ 0x0002 ]
 
 
-%clean
-rm -rf %{buildroot}
-
```
Comment (agentzh) [inline rpm/SPECS/openresty.spec]: Why removing this?

Comment (zxcvbn4038) [inline rpm/SPECS/openresty.spec]: @agentzh In the absence of a "%clean" directory, the default action is to rm -rf buildroot. If you add %clean by itself then buildroot is preserved (good for debugging). It didn't seem necessary to explicitly do the rm -rf.

Comment (agentzh) [inline rpm/SPECS/openresty.spec]: @zxcvbn4038 Since this is automatically generated by the `rpmspec`, I guess it might be required for older versions of rpmbuild. Not sure. I suggest we preserve it (and it is also not related to the current PR's topic).

**Comment by agentzh** [comment]
@zxcvbn4038 At the bottom of the installation section in `openresty-pcre.spec`, you should add the following line:

```
# to silence the check-rpath error
export QA_RPATHS=$[ 0x0002 ]
```

Otherwise the rpmbulid process fails with the following errors at least on Fedora 22:

```
ERROR   0002: file '/usr/local/openresty/pcre/lib64/libpcre.so.1.2.7' contains an invalid rpath '/usr/local/openresty/pcre/lib64' in [/usr/local/openresty/pcre/lib64]
ERROR   0002: file '/usr/local/openresty/pcre/lib64/libpcreposix.so.0.0.4' contains an invalid rpath '/usr/local/openresty/pcre/lib64' in [/usr/local/openresty/pcre/lib64]
ERROR   0002: file '/usr/local/openresty/pcre/lib64/libpcrecpp.so.0.0.1' contains an invalid rpath '/usr/local/openresty/pcre/lib64' in [/usr/local/openresty/pcre/lib64]
```

See the existing `openresty.spec` file for such an example.

I guess you also need that for `openresty-zlib`.

File: rpm/SPECS/openresty-pcre.spec [hunk 43]
```diff
@@ -0,0 +1,62 @@
...
+%description devel
+Development files for Perl-compatible regular expression library for use by OpenResty ONLY
+
+%prep
+%setup -q -n pcre-%{version}
+
+%build
+LDFLAGS="-Wl,-rpath,%{pcre_prefix}/%{_lib}" ./configure \
+  --prefix=%{pcre_prefix} \
+  --libdir=%{pcre_prefix}/%{_lib} \
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Can we get rid of this `--libdir=PATH` option?

Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: We should not use `lib64` on x86_64, for example. Just use the default `lib` directory name should be sufficient.

Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: We do not want to support multi-arch installations for OpenResty anyway.

File: rpm/SPECS/openresty-openssl.spec [hunk 44]
```diff
@@ -38,23 +43,29 @@ Provides C header and static library for OpenResty's OpenSSL library.
...
 %build
-./config --prefix=%{openssl_prefix} \
-    no-threads \
-    shared -g
-
+./config \
+  no-threads shared zlib -g \
+  --openssldir=%{openssl_prefix} \
+  -I%{zlib_prefix}/include \
+  -L%{zlib_prefix}/%{_lib} \
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Let's replace `%{_lib}` with `lib`.

File: rpm/SPECS/openresty-openssl.spec [hunk 45]
```diff
@@ -38,23 +43,29 @@ Provides C header and static library for OpenResty's OpenSSL library.
...
+  -Wl,-rpath,%{zlib_prefix}/%{_lib} \
+  -Wl,-rpath,%{openssl_prefix}/%{_lib}
```
Comment (agentzh) [inline rpm/SPECS/openresty-openssl.spec]: Ditto.

File: rpm/SPECS/openresty.spec [hunk 46]
```diff
@@ -1,37 +1,38 @@
-Name:           openresty
-Version:        1.9.15.1
-Release:        16%{?dist}
-Summary:        OpenResty, scalable web platform by extending NGINX with Lua
+Name:          openresty
+Version:       1.9.15.1
+Release:       16%{?dist}
```
Comment (agentzh) [inline rpm/SPECS/openresty.spec]: Better increment the release number.

File: rpm/SPECS/openresty-pcre.spec [hunk 47]
```diff
@@ -0,0 +1,62 @@
...
+make install DESTDIR=%{buildroot}
+rm -rf %{buildroot}/%{pcre_prefix}/bin
+rm -rf %{buildroot}/%{pcre_prefix}/share
+rm -f  %{buildroot}/%{pcre_prefix}/%{_lib}/*.a
+rm -f  %{buildroot}/%{pcre_prefix}/%{_lib}/*.la
+rm -rf %{buildroot}/%{pcre_prefix}/%{_lib}/pkgconfig
+
+%files
+%defattr(-,root,root,-)
+%{pcre_prefix}/%{_lib}/*.so*
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: We should exclude `libpcrecpp.so*` and `libpcreposix.so*` since we never link against them.

File: rpm/SPECS/openresty-pcre.spec [hunk 48]
```diff
@@ -0,0 +1,62 @@
...
+  --enable-utf \
+  --enable-unicode-properties \
+  --enable-pcre8
+make %{?_smp_mflags}
+
+%install
+make install DESTDIR=%{buildroot}
+rm -rf %{buildroot}/%{pcre_prefix}/bin
+rm -rf %{buildroot}/%{pcre_prefix}/share
+rm -f  %{buildroot}/%{pcre_prefix}/%{_lib}/*.a
```
Comment (agentzh) [inline rpm/SPECS/openresty-pcre.spec]: Better put the `*.a` files into the `devel` sub-package.

File: rpm/SPECS/openresty-zlib.spec [hunk 49]
```diff
@@ -0,0 +1,65 @@
...
+The zlib header files for use by OpenResty ONLY
+
+%prep
+%setup -q -n zlib-%{version}
+
+%build
+export CFLAGS="$RPM_OPT_FLAGS"
+export LDFLAGS="$LDFLAGS -Wl,-rpath,%{zlib_prefix}/%{_lib} -Wl,-z,relro -Wl,-z,now"
+./configure \
+  --libdir=%{zlib_prefix}/%{_lib} \
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: No `--libdir=PATH` option please.

File: rpm/SPECS/openresty-zlib.spec [hunk 50]
```diff
@@ -0,0 +1,65 @@
...
+  --includedir=%{zlib_prefix}/include \
+  --prefix=%{zlib_prefix}
+make %{?_smp_mflags}
+
+%install
+make install DESTDIR=%{buildroot}
+rm -rf %{buildroot}/%{zlib_prefix}/share
+rm -f  %{buildroot}/%{zlib_prefix}/%{_lib}/*.la
+rm -f  %{buildroot}/%{zlib_prefix}/%{_lib}/*.a
```
Comment (agentzh) [inline rpm/SPECS/openresty-zlib.spec]: Better put the `*.a` files into the `devel` sub-package.

**Comment by agentzh** [comment]
@zxcvbn4038 Yes, you also need to set that `QA_RPATHS` environment in the `%install` spec file section in openresty-zlib to make it build on Fedora 22 at least.

**Comment by zxcvbn4038** [comment]
@agentzh Thanks for the feedback, give me a couple days and I'll send an update.

**Comment by agentzh** [comment]
@zxcvbn4038 Thanks a lot! Hopefully we can get this merged before OpenResty 1.11.2.1 is out (right now RC1 has just been out) :)

**Comment by zxcvbn4038** [comment]
@agentzh I'd like that also, its been a busy week but hopefully can make the next set of changes over the weekend.

**Comment by agentzh** [comment]
@zxcvbn4038 Cool, thanks!

File: rpm/SPECS/openresty-openssl.spec (hunk 39 above)
Comment (zxcvbn4038) [inline rpm/SPECS/openresty-openssl.spec]: @agentzh I will try again to use lib, problem is that openssl does magic things to try and decide if files go in lib or lib64.

File: rpm/SPECS/openresty-openssl.spec (hunk 44 above)
Comment (zxcvbn4038) [inline rpm/SPECS/openresty-openssl.spec]: @agentzh Ok, for openssl we need a --libdir to tell it to use "lib" otherwise it seems to be the opposite of wherever you put the other libraries (lib or lib64). I think they were concerned more about it building out of the box for major OS and distributions then people doing special things - but at least they provided an option.

**Comment by zxcvbn4038** [comment]
@agentzh Ok, think I've knocked out everything from your second round of feedback.

File: rpm/SPECS/openresty.spec [hunk 51]
```diff
@@ -163,11 +159,9 @@ mkdir -p %{buildroot}/etc/init.d
 # to silence the check-rpath error
 export QA_RPATHS=$[ 0x0002 ]
 
-
 %clean
 rm -rf %{buildroot}
 
-
```
Comment (agentzh) [inline rpm/SPECS/openresty.spec]: Such blank lines are actually intended. We use 2 blank lines to separate sections like `%clean` and `%post`.

**Comment by agentzh** [comment]
I'm closing this in favor of #2. @zxcvbn4038 Please review #2 for me. I've made a lot of small changes to your patches, some of which are important.
